pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,607 @@
+package awskms
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"github.com/spiffe/spire/pkg/server/plugin/keymanager"
+	"github.com/spiffe/spire/proto/spire/common/plugin"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	pluginName       = "awskms"
+	aliasPrefix      = "alias/"
+	defaultKeyPrefix = "SPIRE_SERVER_KEY/"
+
+	keyIDTag = "key_id"
+	aliasTag = "alias"
+)
+
+func BuiltIn() catalog.Plugin {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.Plugin {
+	return catalog.MakePlugin(pluginName, keymanager.PluginServer(p))
+}
+
+type keyEntry struct {
+	KMSKeyID  string
+	Alias     string
+	PublicKey *keymanager.PublicKey
+}
+
+type pluginHooks struct {
+	newClient func(ctx context.Context, config *Config) (kmsClient, error)
+	// just for testing scheduleDeleteTask
+	deleteSignal chan struct{}
+}
+
+// Plugin is the main representation of this keymanager plugin
+type Plugin struct {
+	keymanager.UnsafeKeyManagerServer
+	log                  hclog.Logger
+	mu                   sync.RWMutex
+	entries              map[string]keyEntry
+	kmsClient            kmsClient
+	keyPrefix            string
+	scheduleDelete       chan string
+	cancelScheduleDelete context.CancelFunc
+	hooks                pluginHooks
+}
+
+// Config provides configuration context for the plugin
+type Config struct {
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	Region          string `hcl:"region" json:"region"`
+	KeyPrefix       string `hcl:"key_prefix" json:"key_prefix"`
+}
+
+// New returns an instantiated plugin
+func New() *Plugin {
+	return newPlugin(newKMSClient)
+}
+
+func newPlugin(newClient func(ctx context.Context, config *Config) (kmsClient, error)) *Plugin {
+	return &Plugin{
+		entries: make(map[string]keyEntry),
+		hooks: pluginHooks{
+			newClient: newClient,
+		},
+		scheduleDelete: make(chan string, 120),
+	}
+}
+
+// SetLogger sets a logger
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure sets up the plugin
+func (p *Plugin) Configure(ctx context.Context, req *plugin.ConfigureRequest) (*plugin.ConfigureResponse, error) {
+	// cancels previous schedule delete task in case of re configure
+	if p.cancelScheduleDelete != nil {
+		p.cancelScheduleDelete()
+	}
+
+	config, err := validateConfig(req.Configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.keyPrefix = config.KeyPrefix
+	p.kmsClient, err = p.hooks.newClient(ctx, config)
```

Comment:
We should avoid mutating the state of the plugin until we're sure the Configure call is going to succeed. In other words, set up everything into local variables and once you are past the point of any chance of error, then and only then mutate the plugin struct members.